internal/provider/maas/environ.go

```diff
@@ -1137,22 +1133,43 @@ func (env *maasEnviron) Spaces(ctx envcontext.ProviderCallContext) (corenetwork.
 	if err != nil {
 		return nil, env.HandleCredentialError(ctx, err)
 	}
+
 	var result []corenetwork.SpaceInfo
+
+	if len(spaces) == 0 {
+		return result, nil
+	}
+
+	// At the time of writing, MAAS zones are effectively just labels.
+	// They do not indicate subnet availability as on say, AWS.
+	// Here, we indicate that all subnets are available in all zones,
+	// and leave it up to the provisioner to see if it can provision
+	// a suitable instance.
+	zones, err := env.maasController.Zones()
+	if err != nil {
+		return nil, env.HandleCredentialError(ctx, err)
+	}
+	zoneNames := transform.Slice(zones, func(zone gomaasapi.Zone) string { return zone.Name() })
+
 	for _, space := range spaces {
-		if len(space.Subnets()) == 0 {
+		subs := space.Subnets()
+		if len(subs) == 0 {
 			continue
 		}
+
 		outSpace := corenetwork.SpaceInfo{
 			Name:       corenetwork.SpaceName(space.Name()),
 			ProviderId: corenetwork.Id(strconv.Itoa(space.ID())),
-			Subnets:    make([]corenetwork.SubnetInfo, len(space.Subnets())),
+			Subnets:    make([]corenetwork.SubnetInfo, len(subs)),
 		}
-		for i, subnet := range space.Subnets() {
+
+		for i, subnet := range subs {
 			subnetInfo := corenetwork.SubnetInfo{
-				ProviderId:      corenetwork.Id(strconv.Itoa(subnet.ID())),
-				VLANTag:         subnet.VLAN().VID(),
-				CIDR:            subnet.CIDR(),
-				ProviderSpaceId: corenetwork.Id(strconv.Itoa(space.ID())),
+				ProviderId:        corenetwork.Id(strconv.Itoa(subnet.ID())),
+				VLANTag:           subnet.VLAN().VID(),
+				CIDR:              subnet.CIDR(),
+				ProviderSpaceId:   corenetwork.Id(strconv.Itoa(space.ID())),
+				AvailabilityZones: zoneNames,
```

Comment:
What happens if we bootstrap with the discovered spaces/subnets and then the availability zones change (or get removed)? Should we update them?

Reply:
But that's the point, we only run it when we create a model or if the user runs `juju reload-spaces`. We are in an inconsistent state until we manually reload them. 
Maybe it's a non-issue.

Reply:
Same with the spaces and subnets themselves.

That's an existing concern though, not one introduced by this patch.

Reply:
Right, fair